src/library/scala/Predef.scala

```diff
@@ -342,6 +342,49 @@ object Predef extends LowPriorityImplicits {
 
   // implicit classes -----------------------------------------------------
 
+  /** Adds `tap` method to every type.
+   *  `tap` applies `f` to the value for its side effects, and returns the original value.
+   *
+   *    {{{
+   *    val xs = List(1, 2, 3)
+   *               .tap(ys => println("debug " + ys.toString))
+   *    // xs == List(1, 2, 3)
+   *    }}}
+   *
+   *  @group implicit-classes-any */
+  implicit final class AnyTap[A](private val self: A) extends AnyVal {
+
+    /** Applies `f` to the value for its side effects, and returns the original value.
+     *
+     *  @param f      the function to apply to the value.
+     *  @tparam U     the result type of the function `f`.
+     *  @return       the original value `self`.
+     */
+    def tap[U](f: A => U): A = { f(self); self }
```

Comment:
- Could the return type be `self.type`?
 - Why the `U` type parameter, not just `A => Unit`?

Reply:
`U` avoids value discarding warnings.

Reply:
Is that desirable? The value (if any) is discarded.

Reply:
IMO yes, as the function is specified as a side-effect.

Reply:
Is there an advantage over having this polymorphic version in `U` over `A => Any`?

Reply:
Not sure, but it looks like there's heaps of precedence in the standard library of doing it this way, see all the definitions of `foreach`.

Reply:
It seems to originate in this PR: https://github.com/scala/scala/commit/ca3d31e7b25e4709839671f2ee5c5bd643cbc08e but the code archaeology ends there. Does it sound plausible that we're cargo-culting this pattern, or are there good reasons for one or the other? Reasons to prefer `Any` are IMO that it's clearer it's side-effecting (it sticks out), and that it's a simpler signature (this signature used to have a `@usecase` annotation, which means it may have been "too complex")

Reply:
I thought the `U` was a convention for `Unit`, but now I see it implies, "needs a `@usecase`".

Reply:
Hot take. Martin likes genericity. The parametric function `A => U` is more beautiful (to me) than `A => Any` though the purists would probably sneer at `tap` as being "lawless."

Reply:
`A => U` implies - to me - a more parametric relation than `A => Any` implies - but the less beautiful thing is actually what you're getting, so I'd argue that should be reflected in the signature.

I don't know who the purists are, why they would consider the tap method lawless, or what lawfulness has to do with anything here in the first place.

Reply:
[Hot take](https://en.wikipedia.org/wiki/Hot_take)

> In journalism, a hot take is a "piece of deliberately provocative commentary that is based almost entirely on shallow moralizing" in response to a news story, "usually written on tight deadlines with little research or reporting, and even less thought".

sounds about right.

By laws, I meant something like Monoid law that can limit the properties of the types. Sorry if I came off as cryptic.

Reply:
> I have the feeling it's a red herring and doesn't have to do with anything.

You could be right, but `tap` is also related to K-combinator `A => B => A`, as seen in [Kestrels](https://github.com/raganwald-deprecated/homoiconic/blob/master/2008-10-29/kestrel.markdown). K can be seen as [`Const`](http://eed3si9n.com/herding-cats/Const.html), which does form a Functor. In all these cases, they probably _can_ be replaced with`Any` (after all it's ignored), but I think `[U]` helps in thinking of `tap` that it has an explicit phantom type.